File: src/classes/containers/sphereContainer.cpp
#include "sphereContainer.hpp"
#include <glm/glm.hpp>

SphereContainer::SphereContainer(glm::vec3 position, glm::vec3 size, bool forcedInside) : Container(position, forcedInside) {
    this->size = size;
}

glm::vec3 SphereContainer::getSize() {
    return size;
}

void SphereContainer::setSize(glm::vec3 size) {
    this->size = size;
}

void SphereContainer::collideWith(Sphere* sphere) {
    // Collision resolution code
    glm::vec3 spherePosition = sphere->position;
    float sphereRadius = sphere->radius;
    float radius = size.x;
    // Calculate the distance between the sphere's position and the container's center
    glm::vec3 axis = spherePosition - position;
    float distance = glm::length(axis);
    
    // Check if the sphere is inside the container
    if (!forcedInside && distance > radius) {
        // the sphere is outside the container
        return;
    }

    // Collision resolution code
    if (distance + sphereRadius > radius) {
        // Calculate the penetration depth
        float penetration = distance + sphereRadius - radius;
        // Calculate the new position of the sphere
        sphere->position -= penetration * glm::normalize(axis);
    }
}

File: rust/src/classes/containers/sphere_container.rs
use nalgebra::{Vector3};

use crate::classes::particles::sphere::Sphere;
use crate::classes::container::{ContainerTrait, Container};

pub struct SphereContainer {
    container: Container,
    size: Vector3<f32>,
}

impl SphereContainer {
    pub fn new(position: Vector3<f32>, size: Vector3<f32>, forced_inside: bool) -> Self {
        SphereContainer {
            container: Container::new(position, forced_inside),
            size,
        }
    }
}

impl ContainerTrait for SphereContainer {
    fn get_position(&self) -> Vector3<f32> {
        self.container.position
    }
    
    fn set_position(&mut self, position: Vector3<f32>) {
        self.container.position = position;
    }
    
    fn get_size(&self) -> Vector3<f32> {
        self.size
    }

    fn set_size(&mut self, size: Vector3<f32>) {
        self.size = size;
    }

    fn collide_with(&self, sphere: &mut Sphere) {
        let sphere_position = sphere.particle.position;
        let sphere_radius = sphere.radius;
        let radius = self.size.x;

        let axis = sphere_position - self.container.position;
        let distance = axis.norm();

        if !self.container.forced_inside && distance > radius {
            // the sphere is outside the container
            return;
        }

        // collision resolution code
        if (distance + sphere_radius) > radius {
            let penetration = distance + sphere_radius - radius;
            sphere.particle.position -= penetration * axis.normalize();
        }
    }
}